### DISCUSSION TITLE
Sandbox based on Vortice and Windows Forms

### DISCUSSION THREAD

**Discussion by MarcioAB** [amerkoleci/Vortice.Windows#369] [General]
I just created a repository named "Sandbox-Vortice-based" on my user ( https://github.com/MarcioAB/Sandbox-Vortice-based )
Can I use this name?
The idea is a sandbox to play with D3D11 and D3D12 based on Vortice and Windows Forms.
So far there are just the essential samples: direct/indexed vertices, instances, MVP transformation, and texture.
This is a work in progress and documentation is in preparation.

**Comment by amerkoleci** [comment]
Hi,
Of course, keep up the good job and keep me updated!

**Comment by MarcioAB** [comment]
Great!

The issue I found so far is related to dispose D3D12 objects.
I'm using Debug Layer ReportLiveObjects to check if any object is still live after I try to dispose of everything.
For D3D11 all objects are properly disposed but for D3D12 there are some objects that remain live (CommandQueue and PipelinesState).
I named all objects but this "extra" CommandQueue has no name. I guess it is created internally.
And despite the PipelineStates are named, I can not dispose them. I think they are somehow referenced by this noname CommandQueue.
This is the ReportLiveObjects after I tried to dispose all (except the device).

*** DXGI: 10 ALL @ D3D_Dispose
   0 Warning    StateCreation        Live ID3D12Device at 0x0000021960ADA368, Name: DEV12_8, Refcount: 8
   1 Warning    StateCreation        	Live ID3D12CommandQueue at 0x000002198454A960, Refcount: 1, IntRef: 0
   2 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219847C0870, Name: Pipeline NDX-0, Refcount: 1, IntRef: 0
   3 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219847C8270, Name: Pipeline NDX-1, Refcount: 1, IntRef: 0
   4 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219847CDC60, Name: Pipeline NDX-2, Refcount: 1, IntRef: 0
   5 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219847A5190, Name: Pipeline NDX-0, Refcount: 1, IntRef: 0
   6 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219847C2B40, Name: Pipeline NDX-1, Refcount: 1, IntRef: 0
   7 Warning    StateCreation        	Live ID3D12PipelineState at 0x00000219846371F0, Name: Pipeline NDX-2, Refcount: 1, IntRef: 0
   8 Warning    StateCreation        Live             ID3D12CommandQueue :      1
   9 Warning    StateCreation        Live            ID3D12PipelineState :      6
The program '[27028] WinFormsApp1.exe' has exited with code 0 (0x0).

Any chance to suggest something I can do to dispose these objects ?

**Reply by amerkoleci** [reply]
Make sure to wait idle for queue to finish pending job before releasing them, follow mine samples here: https://github.com/amerkoleci/Vortice.Windows.Samples/tree/main/src/Direct3D12

**Reply by MarcioAB** [reply]
Correct. I forgot WaitForGpu() in one of many places where I dispose D3D12 objects. Thank you.
